src/core/weights.js

```diff
@@ -165,3 +166,90 @@ export function toJSON(weights: Weights): WeightsJSON {
 export function fromJSON(json: WeightsJSON): Weights {
   return parser.parseOrThrow(json);
 }
+
+export type NodeWeightDiff = {|
+  +address: NodeAddressT,
+  +first: NodeWeight | null,
+  +second: NodeWeight | null,
+|};
+
+export type EdgeWeightDiff = {|
+  +address: EdgeAddressT,
+  +first: EdgeWeight | null,
+  +second: EdgeWeight | null,
+|};
+
+export type WeightsComparison = {|
+  +weightsAreEqual: boolean,
+  +nodeWeightDiffs: $ReadOnlyArray<NodeWeightDiff>,
+  +edgeWeightDiffs: $ReadOnlyArray<EdgeWeightDiff>,
+|};
+
+export function compareWeights(
+  firstWeights: Weights,
+  secondWeights: Weights
+): WeightsComparison {
+  const nodeWeightDiffs = [];
+  const edgeWeightDiffs = [];
+  const weightsAreEqual = deepEqual(firstWeights, secondWeights);
+
+  if (!weightsAreEqual) {
+    for (const [address, firstWeight] of firstWeights.nodeWeights) {
+      const secondWeight = secondWeights.nodeWeights.get(address);
+      if (secondWeight !== undefined) {
+        if (!deepEqual(firstWeight, secondWeight))
+          nodeWeightDiffs.push({
+            address: address,
+            first: firstWeight,
+            second: secondWeight,
+          });
+      } else {
+        nodeWeightDiffs.push({
+          address: address,
+          first: firstWeight,
+          second: null,
+        });
+      }
+    }
```

Comment:
how about:

```js

const addresses = new Set([...firstWeights.nodeWeights.keys(), ...secondWeights.nodeWeights.keys()])
for (const a of addresses) {
  const first = firstWeights.nodeWeights.get(a)
  const second = secondWeights.nodeWeights.get(a)
  if (first !== second) {
    nodeWeightDiffs.push({
      address: a,
      first: first === undefined ? null : first,
     second: second === undefined ? null : second
  })
```

This way you only need one loop for both sides of the diff.

(and likewise for edges)

Note if we want to avoid the `first === undefined ? null : first` bit, we could just change the typedef so instead of being `NodeWeight | null` it's just `?NodeWeight`. Often I like specifying that we'll have `null` not `undefined`, but in this case it's kind of natural since `Map.get(` returns undefined if there's no entry for that key.

Reply:
sure, that's nicer.